src/Adapter/MSTest.TestAdapter/Extensions/MethodInfoExtensions.cs

```diff
@@ -171,55 +171,39 @@ internal static void InvokeAsSynchronousTask(this MethodInfo methodInfo, object?
         {
             int methodParametersLengthOrZero = methodParameters?.Length ?? 0;
             int argumentsLengthOrZero = arguments?.Length ?? 0;
-            if (methodParametersLengthOrZero != argumentsLengthOrZero
-                || !AreArgumentAndParameterTypesAssignable(arguments, methodParameters))
+            if (methodParametersLengthOrZero != argumentsLengthOrZero)
             {
-                throw new TestFailedException(
-                    ObjectModel.UnitTestOutcome.Error,
-                    string.Format(
-                        CultureInfo.InvariantCulture,
-                        Resource.CannotRunTestArgumentsMismatchError,
-                        methodInfo.DeclaringType!.FullName,
-                        methodInfo.Name,
-                        methodParametersLengthOrZero,
-                        string.Join(", ", methodParameters?.Select(p => p.ParameterType.Name) ?? Array.Empty<string>()),
-                        argumentsLengthOrZero,
-                        string.Join(", ", arguments?.Select(a => a?.GetType().Name ?? "null") ?? Array.Empty<string>())));
+                throw GetParameterCountMismatchException(methodInfo, arguments, methodParameters, methodParametersLengthOrZero, argumentsLengthOrZero, innerException: null);
             }
 
-            task = methodInfo.Invoke(classInstance, arguments) as Task;
+            try
+            {
+                task = methodInfo.Invoke(classInstance, arguments) as Task;
+            }
+            catch (TargetParameterCountException ex)
+            {
+                throw GetParameterCountMismatchException(methodInfo, arguments, methodParameters, methodParametersLengthOrZero, argumentsLengthOrZero, ex);
+            }
+            catch (ArgumentException ex)
```

Comment:
Nit using the catch/when clause.

```suggestion
            catch (Exception ex) when (ex is TargetParameterCountException or ArgumentException)
```